File: Famicoco/app/src/main/java/com/example/mito/famicoco/MainActivity.java
package com.example.mito.famicoco;

import android.annotation.SuppressLint;
import android.bluetooth.BluetoothAdapter;
import android.bluetooth.BluetoothManager;
import android.bluetooth.le.BluetoothLeScanner;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.SharedPreferences;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.AsyncTask;
import android.os.Bundle;
import android.os.Handler;
import android.preference.PreferenceManager;
import android.support.design.widget.TabLayout;
import android.support.v4.app.AppLaunchChecker;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentPagerAdapter;
import android.support.v4.view.ViewPager;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import org.json.JSONArray;
import org.json.JSONException;

import java.io.IOException;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.HashMap;
import java.util.Timer;
import java.util.TimerTask;

import butterknife.BindView;
import butterknife.ButterKnife;
import okhttp3.FormBody;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;

import static com.example.mito.famicoco.IBeaconScanCallback.beaconId;
import static com.example.mito.famicoco.IBeaconScanCallback.f_init;

public class MainActivity extends AppCompatActivity {

    BluetoothLeScanner mBluetoothLeScanner;
    public static String myBeaconMacAddress = "";
    IBeaconScanCallback mScanCallback;
    public static final String ServerUrl = "SERVERURL";
    public static String myName = "";
    public static Bitmap tlIcon[] = new Bitmap[6];
    public static String myRegistrationId = "";
    private Handler mHandler = null;
    public static HashMap<String, Bitmap> selectIcon_map = null;
    public static HashMap<String, Integer> selectIcon_map_id = null;

    @BindView(R.id.tabs)
    TabLayout tabLayout;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        ButterKnife.bind(this);

        final BluetoothManager bManager = (BluetoothManager) getSystemService(Context.BLUETOOTH_SERVICE);
        assert bManager != null;
        final BluetoothAdapter bluetoothAdapter = bManager.getAdapter();

        setTitle(getString(R.string.tl));

        selectIcon_map = new HashMap<>();
        selectIcon_map_id = new HashMap<>();
        tlIcon[0] = BitmapFactory.decodeResource(getResources(), R.drawable.setting);
        tlIcon[1] = BitmapFactory.decodeResource(getResources(), R.drawable.home);
        tlIcon[2] = BitmapFactory.decodeResource(getResources(), R.drawable.car);
        tlIcon[3] = BitmapFactory.decodeResource(getResources(), R.drawable.humen);
        tlIcon[4] = tlIcon[3];
        tlIcon[5] = BitmapFactory.decodeResource(getResources(), R.drawable.pet);

        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        // Create the adapter that will return a fragment for each of the three
        // primary sections of the activity.
        /*
      The {@link android.support.v4.view.PagerAdapter} that will provide
      fragments for each of the sections. We use a
      {@link FragmentPagerAdapter} derivative, which will keep every
      loaded fragment in memory. If this becomes too memory intensive, it
      may be best to switch to a
      {@link android.support.v4.app.FragmentStatePagerAdapter}.
     */
        SectionsPagerAdapter mSectionsPagerAdapter = new SectionsPagerAdapter(getSupportFragmentManager());

        // Set up the ViewPager with the sections adapter.
        /*
      The {@link ViewPager} that will host the section contents.
     */
        ViewPager mViewPager = (ViewPager) findViewById(R.id.container);
        mViewPager.setAdapter(mSectionsPagerAdapter);

        mViewPager.addOnPageChangeListener(new ViewPager.OnPageChangeListener() {
            @Override
            public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
            }

            @Override
            public void onPageSelected(int position) {
                KeyboardUtils.hide(MainActivity.this);
                if (position == 0)
                    setTitle(getString(R.string.tl));
                else if (position == 1)
                    setTitle(getString(R.string.ie));
                else if (position == 2)
                    setTitle(getString(R.string.soto));
                else
                    setTitle(getString(R.string.pet));
            }

            @Override
            public void onPageScrollStateChanged(int state) {
            }
        });

        tabLayout.setupWithViewPager(mViewPager);

        TabLayout.Tab tab;
        int[] drawableId = new int[]{R.drawable.tab1_selector, R.drawable.tab2_selector, R.drawable.tab3_selector, R.drawable.tab4_selector};
        for (int i = 0; i < 4; i++) {
            tab = tabLayout.getTabAt(i);
            assert tab != null;
            tab.setIcon(drawableId[i]);
        }

        bluetoothAdapter.enable(); //OK

        if (!bluetoothAdapter.isEnabled())
            new AlertDialog.Builder(this)
                    .setTitle("確認")
                    .setMessage("Bluetooth機能をONにしました")
                    .setPositiveButton("OK", new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialogInterface, int i) {
                        }
                    })
                    .show();

        while (!bluetoothAdapter.isEnabled()) ;

        // BLEScan開始 not Estimote sdk
        mBluetoothLeScanner = bluetoothAdapter.getBluetoothLeScanner();
        mScanCallback = new IBeaconScanCallback();
        mBluetoothLeScanner.startScan(mScanCallback);

        Log.d("before", "log");
        if (AppLaunchChecker.hasStartedFromLauncher(this)) {
            Log.d("AppLaunchChecker", "2回目以降");
            SharedPreferences sp = PreferenceManager.getDefaultSharedPreferences(this);
            myName = sp.getString("name", null);
            myBeaconMacAddress = sp.getString("myBeacon", null);
            f_init = false;
        } else {
            Log.d("AppLaunchChecker", "はじめてアプリを起動した");
            Intent intent = new Intent(this, InitStartActivity.class);
            startActivity(intent);
            AppLaunchChecker.onActivityCreate(this);
        }
        connect();
    }

    public void connect() {
        mHandler = new Handler();
        Timer mTimer = new Timer();
        mTimer.schedule(new TimerTask() {
            @Override
            public void run() {
                mHandler.post(new Runnable() {
                    @Override
                    public void run() {
                        // 実行したい処理
                        @SuppressLint("StaticFieldLeak") AsyncTask<URL, Void, Boolean> asyncTask = new AsyncTask<URL, Void, Boolean>() {
                            @Override
                            protected Boolean doInBackground(URL... urls) {
                                final URL url = urls[0];
                                HttpURLConnection con;

                                try {
                                    con = (HttpURLConnection) url.openConnection();
                                    con.setDoInput(true);
                                    Log.d("TAG", "Connected");
                                    con.connect();
                                    InputStream in = con.getInputStream();
                                    String readSt = HttpGetTask.readInputStream(in);

                                    //// 配列を取得する場合
                                    JSONArray jsonArray = new JSONArray(readSt);
                                    int m = jsonArray.length();
                                    for (int i = 0; i < m; i++) {
                                        JSONArray array = jsonArray.getJSONArray(i);
                                        String num = array.getString(1);
                                        String str = array.getString(0);
                                        selectIcon_map_id.put(str, num.charAt(0) - '0');
                                    }

                                } catch (IOException | JSONException e) {
                                    e.printStackTrace();
                                }
                                return true;
                            }
                        };

                        try {
                            URL url = new URL(ServerUrl + "user/pict/");
                            asyncTask.execute(url);
                        } catch (MalformedURLException e) {
                            e.printStackTrace();
                        }

                        @SuppressLint("StaticFieldLeak") AsyncTask<URL, Void, Boolean> task = new AsyncTask<URL, Void, Boolean>() {
                            @Override
                            protected Boolean doInBackground(URL... urls) {
                                final URL url = urls[0];

                                OkHttpClient client = new OkHttpClient();
                                JSONArray array = new JSONArray();
                                for (int i = 0; i < beaconId.size(); i++) {
                                    array.put(beaconId.get(i));
                                }

                                RequestBody body = new FormBody.Builder()
                                        .add("data", array.toString())
                                        .add("sender", myName)
                                        .build();
                                Request request = new Request.Builder()
                                        .url(url)
                                        .post(body)
                                        .build();

                                try {
                                    client.newCall(request).execute();
                                } catch (IOException e) {
                                    e.printStackTrace();
                                }

                                beaconId.clear();
                                return true;
                            }
                        };

                        try {
                            URL url = new URL(ServerUrl + "receive_beacon/");
                            task.execute(url);
                        } catch (MalformedURLException e) {
                            e.printStackTrace();
                        }
                    }
                });
            }
        }, 5000, 5000);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        int id = item.getItemId();

        Intent intent = null;
        if (id == R.id.notification_setting) {
            intent = new Intent(getApplicationContext(), NotificationActivity.class);
        } else if (id == R.id.pet_setting) {
            intent = new Intent(getApplicationContext(), PetSettingActivity.class);
        }
        startActivity(intent);
        return super.onOptionsItemSelected(item);
    }

    /**
     * A placeholder fragment containing a simple view.
     */
    public static class PlaceholderFragment extends Fragment {
        /**
         * The fragment argument representing the section number for this
         * fragment.
         */
        private static final String ARG_SECTION_NUMBER = "section_number";

        public PlaceholderFragment() {
        }


        @Override
        public View onCreateView(LayoutInflater inflater, ViewGroup container,
                                 Bundle savedInstanceState) {
            View rootView = inflater.inflate(R.layout.fragment_main, container, false);
            TextView textView = (TextView) rootView.findViewById(R.id.section_label);
            textView.setText(getString(R.string.section_format, getArguments().getInt(ARG_SECTION_NUMBER)));
            return rootView;
        }
    }

    /**
     * A {@link FragmentPagerAdapter} that returns a fragment corresponding to
     * one of the sections/tabs/pages.
     */
    public class SectionsPagerAdapter extends FragmentPagerAdapter {

        SectionsPagerAdapter(FragmentManager fm) {
            super(fm);
        }

        //タブのフラグメントを渡す
        @Override
        public Fragment getItem(int position) {
            // getItem is called to instantiate the fragment for the given page.
            // Return a PlaceholderFragment (defined as a static inner class below).
            switch (position) {
                case 0:
                    return new TLFragment();
                case 1:
                    return new IEFragment();
                case 2:
                    return new SOTOFragment();
                case 3:
                    return new PETFragment();
            }
            return null;
        }

        //タブの数
        @Override
        public int getCount() {
            // Show 3 total pages.
            return 4;
        }

        //タブの名前
        @Override
        public CharSequence getPageTitle(int position) {
            return null;
        }
    }
}

File: Famicoco/app/src/main/java/com/example/mito/famicoco/MainActivity.kt
package com.example.mito.famicoco

import android.annotation.SuppressLint
import android.bluetooth.BluetoothManager
import android.bluetooth.le.BluetoothLeScanner
import android.content.Context
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.os.AsyncTask
import android.os.Bundle
import android.os.Handler
import android.preference.PreferenceManager
import android.util.Log
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.core.app.AppLaunchChecker
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentPagerAdapter
import androidx.viewpager.widget.ViewPager
import butterknife.BindView
import butterknife.ButterKnife
import com.example.mito.famicoco.IBeaconScanCallback.Companion.beaconIds
import com.example.mito.famicoco.IBeaconScanCallback.Companion.f_init
import com.google.android.material.tabs.TabLayout
import okhttp3.FormBody
import okhttp3.OkHttpClient
import okhttp3.Request
import org.json.JSONArray
import java.io.IOException
import java.net.HttpURLConnection
import java.net.MalformedURLException
import java.net.URL
import java.util.HashMap
import java.util.Timer
import java.util.TimerTask

class MainActivity : AppCompatActivity() {
    
    private lateinit var mBluetoothLeScanner: BluetoothLeScanner
    private lateinit var mScanCallback: IBeaconScanCallback
    private var mHandler: Handler? = null
    
    @BindView(R.id.tabs)
    internal lateinit var tabLayout: TabLayout
    
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        ButterKnife.bind(this)
        
        val bManager = getSystemService(Context.BLUETOOTH_SERVICE) as BluetoothManager
        val bluetoothAdapter = bManager.adapter
        
        title = getString(R.string.tl)
        
        tlIcon[0] = BitmapFactory.decodeResource(
                resources,
                R.drawable.setting
        )
        tlIcon[1] = BitmapFactory.decodeResource(
                resources,
                R.drawable.home
        )
        tlIcon[2] = BitmapFactory.decodeResource(
                resources,
                R.drawable.car
        )
        tlIcon[3] = BitmapFactory.decodeResource(
                resources,
                R.drawable.humen
        )
        tlIcon[4] = tlIcon[3]
        tlIcon[5] = BitmapFactory.decodeResource(
                resources,
                R.drawable.pet
        )
        
        val toolbar = findViewById<View>(R.id.toolbar) as Toolbar
        setSupportActionBar(toolbar)
        // Create the adapter that will return a fragment for each of the three
        // primary sections of the activity.
        /*
      The {@link android.support.v4.view.PagerAdapter} that will provide
      fragments for each of the sections. We use a
      {@link FragmentPagerAdapter} derivative, which will keep every
      loaded fragment in memory. If this becomes too memory intensive, it
      may be best to switch to a
      {@link android.support.v4.app.FragmentStatePagerAdapter}.
     */
        val mSectionsPagerAdapter = SectionsPagerAdapter(supportFragmentManager)
        
        // Set up the ViewPager with the sections adapter.
        /*
      The {@link ViewPager} that will host the section contents.
     */
        val mViewPager = findViewById<View>(R.id.container) as ViewPager
        mViewPager.adapter = mSectionsPagerAdapter
        
        mViewPager.addOnPageChangeListener(object : ViewPager.OnPageChangeListener {
            override fun onPageScrolled(
                    position: Int,
                    positionOffset: Float,
                    positionOffsetPixels: Int
            ) {
            }
            
            override fun onPageSelected(position: Int) {
                KeyboardUtils.hide(this@MainActivity)
                title = when (position) {
                    0 -> getString(R.string.tl)
                    1 -> getString(R.string.ie)
                    2 -> getString(R.string.soto)
                    else -> getString(R.string.pet)
                }
            }
            
            override fun onPageScrollStateChanged(state: Int) {}
        })
        
        tabLayout.setupWithViewPager(mViewPager)
        
        var tab: TabLayout.Tab?
        val drawableId = intArrayOf(
                R.drawable.tab1_selector,
                R.drawable.tab2_selector,
                R.drawable.tab3_selector,
                R.drawable.tab4_selector
        )
        for (i in 0..3) {
            tab = tabLayout.getTabAt(i)
            assert(tab != null)
            tab!!.setIcon(drawableId[i])
        }
        
        bluetoothAdapter.enable() // OK
        
        if (!bluetoothAdapter.isEnabled)
            AlertDialog.Builder(this)
                    .setTitle("確認")
                    .setMessage("Bluetooth機能をONにしました")
                    .setPositiveButton("OK") { _, _ -> }
                    .show()
        
        while (!bluetoothAdapter.isEnabled)
        
        // BLEScan開始 not Estimote sdk
            mBluetoothLeScanner = bluetoothAdapter.bluetoothLeScanner
        mScanCallback = IBeaconScanCallback()
        mBluetoothLeScanner.startScan(mScanCallback)
        
        if (AppLaunchChecker.hasStartedFromLauncher(this)) {
            Log.d("AppLaunchChecker", "2回目以降")
            val sp = PreferenceManager.getDefaultSharedPreferences(this)
            myName = sp.getString(
                    "name",
                    null
            ) ?: ""
            myBeaconMacAddress = sp.getString(
                    "myBeacon",
                    null
            ) ?: ""
            f_init = false
        } else {
            Log.d("AppLaunchChecker", "はじめてアプリを起動した")
            val intent = Intent(
                    this,
                    InitStartActivity::class.java
            )
            startActivity(intent)
            AppLaunchChecker.onActivityCreate(this)
        }
        connect()
    }
    
    private fun connect() {
        mHandler = Handler()
        val mTimer = Timer()
        mTimer.schedule(
                object : TimerTask() {
                    override fun run() {
                        mHandler!!.post {
                            // 実行したい処理
                            @SuppressLint("StaticFieldLeak")
                            val asyncTask = object : AsyncTask<URL, Void, Boolean>() {
                                override fun doInBackground(vararg urls: URL): Boolean {
                                    val url = urls[0]
                                    val con: HttpURLConnection
                                    
                                    try {
                                        con = url.openConnection() as HttpURLConnection
                                        con.doInput = true
                                        con.connect()
                                        val `in` = con.inputStream
                                        val readSt = HttpGetTask.readInputStream(`in`)
                                        
                                        // // 配列を取得する場合
                                        val jsonArray = JSONArray(readSt)
                                        val m = jsonArray.length()
                                        for (i in 0 until m) {
                                            val array = jsonArray.getJSONArray(i)
                                            val num = array.getString(1)
                                            val str = array.getString(0)
                                            selectIconMapId[str] = num[0] - '0'
                                        }
                                    } catch (e: Exception) {
                                        e.printStackTrace()
                                    }
                                    
                                    return true
                                }
                            }
                            
                            try {
                                val url = URL(ServerUrl + "user/pict/")
                                asyncTask.execute(url)
                            } catch (e: MalformedURLException) {
                                e.printStackTrace()
                            }
                            
                            @SuppressLint("StaticFieldLeak")
                            val task = object : AsyncTask<URL, Void, Boolean>() {
                                override fun doInBackground(vararg urls: URL): Boolean {
                                    val url = urls[0]
                                    
                                    val client = OkHttpClient()
                                    val array = JSONArray()
                                    for (i in beaconIds.indices) {
                                        array.put(beaconIds[i])
                                    }
                                    
                                    val body = FormBody.Builder()
                                            .add(
                                                    "data",
                                                    array.toString()
                                            )
                                            .add(
                                                    "sender",
                                                    myName
                                            )
                                            .build()
                                    val request = Request.Builder()
                                            .url(url)
                                            .post(body)
                                            .build()
                                    
                                    try {
                                        client.newCall(request).execute()
                                    } catch (e: IOException) {
                                        e.printStackTrace()
                                    }
                                    
                                    beaconIds.clear()
                                    return true
                                }
                            }
                            
                            try {
                                val url = URL(ServerUrl + "receive_beacon/")
                                task.execute(url)
                            } catch (e: MalformedURLException) {
                                e.printStackTrace()
                            }
                        }
                    }
                },
                5000,
                5000
        )
    }
    
    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        // Inflate the menu; this adds items to the action bar if it is present.
        menuInflater.inflate(
                R.menu.menu_main,
                menu
        )
        return true
    }
    
    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        val id = item.itemId
        
        var intent: Intent? = null
        when (id) {
            R.id.notification_setting -> intent = Intent(
                    applicationContext,
                    NotificationActivity::class.java
            )
            R.id.pet_setting -> intent = Intent(
                    applicationContext,
                    PetSettingActivity::class.java
            )
        }
        startActivity(intent)
        return super.onOptionsItemSelected(item)
    }
    
    /**
     * A placeholder fragment containing a simple view.
     */
    class PlaceholderFragment : Fragment() {
        
        override fun onCreateView(
                inflater: LayoutInflater,
                container: ViewGroup?,
                savedInstanceState: Bundle?
        ): View? {
            val rootView = inflater.inflate(
                    R.layout.fragment_main,
                    container,
                    false
            )
            val textView = rootView.findViewById<View>(R.id.section_label) as TextView
            textView.text = getString(R.string.section_format, arguments!!.getInt(ARG_SECTION_NUMBER))
            return rootView
        }
        
        companion object {
            /**
             * The fragment argument representing the section number for this
             * fragment.
             */
            private const val ARG_SECTION_NUMBER = "section_number"
        }
    }
    
    /**
     * A [FragmentPagerAdapter] that returns a fragment corresponding to
     * one of the sections/tabs/pages.
     */
    inner class SectionsPagerAdapter internal constructor(fm: FragmentManager) : FragmentPagerAdapter(fm) {
        
        // タブのフラグメントを渡す
        override fun getItem(position: Int): Fragment? {
            // getItem is called to instantiate the fragment for the given page.
            // Return a PlaceholderFragment (defined as a static inner class below).
            when (position) {
                0 -> return TLFragment()
                1 -> return IeFragment()
                2 -> return SotoFragment()
                3 -> return PetFragment()
            }
            return null
        }
        
        // タブの数
        override fun getCount(): Int {
            // Show 3 total pages.
            return 4
        }
        
        // タブの名前
        override fun getPageTitle(position: Int): CharSequence? {
            return null
        }
    }
    
    companion object {
        var myBeaconMacAddress: String = ""
        const val ServerUrl = "SERVERURL"
        var myName: String = ""
        var tlIcon = ArrayList<Bitmap>(6)
        var myRegistrationId = ""
        var selectIconMap: HashMap<String, Bitmap> = HashMap()
        var selectIconMapId: HashMap<String, Int> = HashMap()
    }
}